File: sdl2-cpp-space-invaders-master/Explosion.cpp
#include "Explosion.h"
#include "Enemy.h"
#include <boost/algorithm/clamp.hpp>


Explosion::Explosion()
{;
	LoadResources();
	spriteRect = { 0, 0, 128, 128 };
}

Explosion::Explosion(SDL_Rect position, int rowid, int maxShootTime)
{;
	LoadResources();
	spriteRect = { 0, 0, 128, 128 };
	pos = position;
}

Explosion& Explosion::operator=(const Explosion& other)
{
	pos = other.pos;
	return *this;
}

Explosion::~Explosion()
{
}

SDL_Rect Explosion::TileBoundingBox()
{
	return { pos.x, (int)pos.y, 128, 128 };
}

void Explosion::setPosition(SDL_Rect position)
{
	pos = position;
}


SDL_Rect Explosion::getPosition()
{
	return pos;
}

void Explosion::LoadResources()
{
	tex2d.Load("rd/explode.png");
}

void Explosion::Animator(float deltaTime)
{
	currentFrame += animationSpeed * deltaTime;

	currentFrame = boost::algorithm::clamp(currentFrame, 0, 16);

	if (currentFrame >= 16)
	{
		//currentFrame = 0;
		animationCompleted = true;
	}

	spriteRect = { (int)currentFrame * 128, 0, 128, 128 };
}

void Explosion::Update(float deltaTime)
{
	Animator(deltaTime);
}

File: libgdx_space_invaders/core/src/com/libgdx/space_invaders/Explosion.java
package com.libgdx.space_invaders;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;

public class Explosion {

    Texture sprite;
    Vector2 pos;
    Rectangle spriteRect = new Rectangle(0, 0, 128, 128);
    Rectangle animationRect;

    private float animationSpeed = 25;
    private float currentFrame;
    public boolean animationCompleted = false;

    Explosion(Vector2 _pos) {
        Create(_pos);
    }

    public Rectangle TileBoundingBox() {
        return new Rectangle(
                pos.x,
                pos.y,
                spriteRect.width,
                spriteRect.height);
    }

    public void Create(Vector2 _pos) {
        sprite = new Texture("explode.png");
        pos = new Vector2(_pos.x, _pos.y);
        spriteRect = new Rectangle(0, 0, sprite.getWidth(), sprite.getHeight());
    }

    private void Animator() {

        currentFrame += animationSpeed * Gdx.graphics.getDeltaTime();
        currentFrame = MathUtils.clamp(currentFrame, 0, 16);

        if (currentFrame >= 16) {
            animationCompleted = true;
        }

        animationRect = new Rectangle((int) currentFrame * 128, 0, 128, 128);
    }


    public void Update() {

        Animator();
        spriteRect = new Rectangle(pos.x, pos.y, 128, 128);
    }

    public void Draw(SpriteBatch batch) {
        batch.draw(sprite, pos.x, pos.y, (int) animationRect.x, (int) animationRect.y, (int) animationRect.width, (int) animationRect.height);
    }
}
